ocf_data_sampler/torch_datasets/process_and_combine.py

```diff
@@ -0,0 +1,143 @@
+
+        for nwp_key, da_nwp in dataset_dict["nwp"].items():
+            # Standardise
+            provider = config.input_data.nwp[nwp_key].nwp_provider
+            da_nwp = (da_nwp - NWP_MEANS[provider]) / NWP_STDS[provider]
+            # Convert to NumpyBatch
+            nwp_numpy_modalities[nwp_key] = convert_nwp_to_numpy_batch(da_nwp)
+
+        # Combine the NWPs into NumpyBatch
+        numpy_modalities.append({NWPBatchKey.nwp: nwp_numpy_modalities})
+
+    if "sat" in dataset_dict:
+        # Satellite is already in the range [0-1] so no need to standardise
+        da_sat = dataset_dict["sat"]
+
+        # Convert to NumpyBatch
+        numpy_modalities.append(convert_satellite_to_numpy_batch(da_sat))
+
+    gsp_config = config.input_data.gsp
+
+    if "gsp" in dataset_dict:
+        da_gsp = xr.concat([dataset_dict["gsp"], dataset_dict["gsp_future"]], dim="time_utc")
+        da_gsp = da_gsp / da_gsp.effective_capacity_mwp
+
+        numpy_modalities.append(
+            convert_gsp_to_numpy_batch(
+                da_gsp, t0_idx=gsp_config.history_minutes / gsp_config.time_resolution_minutes
+            )
+        )
+
+        # Add coordinate data
+        # TODO: Do we need all of these?
+        numpy_modalities.append(
+            {
+                GSPBatchKey.gsp_id: location.id,
+                GSPBatchKey.x_osgb: location.x,
+                GSPBatchKey.y_osgb: location.y,
+            }
+        )
+
+
+    if "site" in dataset_dict:
+        site_config = config.input_data.site
+        da_sites = dataset_dict["site"]
+        da_sites = da_sites / da_sites.capacity_kwp
+
+        numpy_modalities.append(
+            convert_site_to_numpy_batch(
+                da_sites, t0_idx=site_config.history_minutes / site_config.time_resolution_minutes
+            )
+        )
+
+    if sun_position_key == 'gsp':
+        # Make sun coords NumpyBatch
+        datetimes = pd.date_range(
+            t0 - minutes(gsp_config.history_minutes),
+            t0 + minutes(gsp_config.forecast_minutes),
+            freq=minutes(gsp_config.time_resolution_minutes),
+        )
+
+        lon, lat = osgb_to_lon_lat(location.x, location.y)
+        key_prefix = "gsp"
+
+    elif sun_position_key == 'site':
+        # Make sun coords NumpyBatch
+        datetimes = pd.date_range(
+            t0 - minutes(site_config.history_minutes),
+            t0 + minutes(site_config.forecast_minutes),
+            freq=minutes(site_config.time_resolution_minutes),
+        )
+
+        lon, lat = location.x, location.y
+        key_prefix = "site"
+
+    numpy_modalities.append(
+        make_sun_position_numpy_batch(datetimes, lon, lat, key_prefix=key_prefix)
+    )
+
+    # Combine all the modalities and fill NaNs
+    combined_sample = merge_dicts(numpy_modalities)
+    combined_sample = fill_nans_in_arrays(combined_sample)
+
+    return combined_sample
+
+
+def merge_dicts(list_of_dicts: list[dict]) -> dict:
```

Comment:
Definitely not for this PR, but out of curiosity: why do we merge dicts separately instead of storing modalities in a dict right away?

Reply:
Ive mentioned it in https://github.com/openclimatefix/ocf-data-sampler/issues/75